wg-api/spec-documents/better-native-window-open.md

```diff
@@ -0,0 +1,94 @@
+## Background
+### Impetus
+
+In the current implementation, we destroy the newly-created `WebContents` when the consumer calls `event.preventDefault()`. In recent Electron versions, this can cause a crash, because the WebContents is still initializing when we kill it. Additionally, the preload script has started to run, which has the potential to cause quite a bit of confusion for developers looking at logs. We could patch and continue to patch these weirdnesses, but fixing the root seems appropriate sooner rather than later.
+
+### Future of window.open
+
+As we align more with Chrome's process model, `BrowserWindowProxy` will become less necessary. `nativeWindowOpen: true` will become the default and the ergonomics (and durability) of this API will matter quite a bit more!
+
+### Current API pain points
+
+1. Returning a window via `event.newGuest` in the event handler is unprecedented in Electron and doesn't fit normal "Web Events" semantics
+2. The `options` argument of `new-window` is a mix of parsed options from the feature string and important fields provided by Electron, like `webContents` itself. It's not clear that you need to pass `webContents` into the overridden window for things to work.
+3. Can't set `webPreferences`, since the webContents was already created
+
+### Use cases
+
+1. Custom handling for target="_blank" links from trusted or untrusted content.
+    - Canceling disallowed urls
+    - Kicking out to a browser vs. opening windows with Electron
+2. Opening application windows with synchronous DOM access
+3. Navigating to external sites (for OAuth flows, for example)
+
+### Platforms
+
+All platforms
+
+## API Design
+
+**Instance method `webContents.setWindowOpenOverride(handler)`**
+
+- `handler` Function
+    - `windowOpenInfo` Object
+        - `url` String
+        - `frameName` String
+        - `disposition` String
+        - `parsedOptions` BrowserWindowConstructorOptions
+        - `additionalFeatures` String[]
+
+    Returns `Boolean | BrowserWindowConstructorOptions`
```

Comment:
Hmm, I don't love returning the constructor options here, that's kind of a weird API. It seems more natural that if you override this function, you're responsible for creating the new window yourself. i.e., the default behavior will be to cancel the new window, and you have to do something to make that not the case. i.e. if you want to open a window as if you hadn't overridden this, you'd do:

```
webContents.setWindowOpenOverride((info) => {
  new BrowserWindow(info.parsedOptions)
})
```

And if we do things that way, it seems sort of odd not to have this be a regular event with an event handler that you'd call `event.preventDefault()` on, no?

Reply:
We can't construct a new window at this point as the webcontents hasn't been made yet.  The window can't be made until chromium has handed us the webcontents.

Reply:
It is weird, but it's the least weird we could come up with, given the constraints that Sam mentioned. The API you describe is basically the "ideal" version I laid out in the original writeup. We came to the reasonable consensus that it was too hard for the moment.